### DISCUSSION TITLE
Periodic Tesselations with Flat Surfaces

### DISCUSSION THREAD

**Discussion by lkuna24** [neperfepx/neper#100] [Q&A] [answered]
To create a tesselation/mesh such that the resulting structure is still a square/cube with flat surface but has periodicity.  To elaborate, the +z face of the cube it would be exactly the same as the -z face, and for all the sides etc.
Resulting cube would have tesselations periodic across all faces.

**Comment by rquey** [comment, marked as answer]
The reason why non-cubic periodic tessellations have been preferred is that cells are not cut, and therefore there is no additional surfaces, etc.  This is easier for generation, regularization (although it is not implemented yet) and meshing. Moreover, and from a general point of view, this should not cause problems from the solver side - the loading just has to be defined a little differently.

**Reply by xinliu09** [reply]
I guess that a cube domain is easier to handle in some micromechanics models, so is there a way to generate a cube domain with periodic mesh in Neper?
Thank you!

**Reply by rquey** [reply]
It may seem easier, but I think it is not the right thing to do...

**Reply by xinliu09** [reply]
Thanks a lot for the prompt responses. I understand the rationale of the current design in generating periodic mesh, but I don't understand a cubic domain is "not the right thing to do". I saw many papers using neper or other software to create a cubic domain with periodic mesh but not periodic grains. It looks like to me a cubic domain with a true periodicity (both for grains and mesh) is very attractive. Please correct me if I am wrong. Thank you.

**Reply by yguilhem** [reply]
@rquey is right because "cutting" the topology of the grains with the planes corresponding to the cube limits will "artificially" add supplementary small edges. That would not be a problem if you are able to mesh (and compute) with a very small element size (smaller than the small edges emanating from the cut). Otherwise you will suffer from potential flat elements and/or varying element size.
Anyway it will for sure be less efficient than the current periodic tesselation method.

**Reply by xinliu09** [reply]
@yguilhem Thank you for the answer. It makes sense to me if considering the mesh quality and efficiency.

**Reply by yguilhem** [reply]
As a complement to my answer, small edges/faces emanating from the cut could vanish by applying regularization (not implemented for periodic tessellations). Nevertheless, regularization may lead to face flatness fault which is another source of difficulties for mesh generation.
Especially if too many small edges are identified by this process and have to be fused into a single vertex (and small faces fused into a single edge).

**Reply by xinliu09** [reply]
@yguilhem Thank you for the clarification. I was thinking about the voxel mesh approach. I know voxel mesh has been used in many RVEs with complex geometries. The good part of voxel mesh is that the mesh quality can be controlled, but it will cause the inaccuracy in the real geometries and maybe unrealistic stress along the interfaces. 
Looks like the current design in Neper is the best one, it is just not very intuitive at first glance.

**Reply by rquey** [reply]
@xinliu09, note that even with a non-cubic periodic tessellation, the "voxel mesh" (`-elt hex`) will be cubic.

**Comment by RoadToDream** [comment]
May I ask why regularization is not implemented for periodicity case? Is there any concerns for implementation or it's just due to the schedule. I would love to help to implement this if there will be some guidance provided.

**Comment by dilekguzel** [comment]
Hi everyone,
To obtain Neper-based periodic microstructures on unit domains (with flat surfaces), we have developed an open-source software, [Neper-Mosaic](https://github.com/InstituteOfMechanics/Neper_Mosaic_RVE_Generator). 
The goal is to use Neper-based tesselations and do the relevant cutting operations so that we generate a unit square/cubic periodic RVEs/microstructures automatically.

**Learn more:**
Code and Documentation: [Neper-Mosaic](https://github.com/InstituteOfMechanics/Neper_Mosaic_RVE_Generator)
Detailed Explanation: [Publication](https://www.sciencedirect.com/science/article/pii/S2352711024002826)

**Reply by MROHOMAN** [reply]
Hi @dilekguzel,

I was playing with your code. Do you have a way to do the same thing for hexahedral elements?

**Reply by dilekguzel** [reply]
Hi @MROHOMAN,
unfortunately obtaining hexahedral elements in such complex geometries like grains are not really feasible, and Mosaic cannot handle that automatically at the moment. However Mosaic is capable to create a new .geo file with the new geometry definitions. One can use that file and manually modify (recombine/extrude) and create mesh via Gmsh (if possible).

Good luck, it is not an easy task :)

**Reply by MROHOMAN** [reply]
@dilekguzel,
Thanks for getting back. I would appreciate if you could answer the following issues,
1. Is there a way to control the number of elements and get the orientation of each element in the MOSAIC? I know, I can get them from NEPER meshing module, but I can only input the tessellation/.tess file in the MOSAIC.

2. How can I have the grains colored in the Abaqus? I imported the file generated by MOSAIC in Abaqus but did not see the grains.

Thanks in advance for your help.

**Reply by dilekguzel** [reply]
Hello, 

1. The short answer is in Mosaic, there is an option -cl (--element_size), which is the characteristic length, similar to Neper. You can specify this number as an option, for example:
python3 -m neper_mosaic n20-id3.geo output/testout.geo_unrolled -v -cl 0.05 --show_gui --show_gmsh
According to given number you can control the number of elements. (The long answer is Mosaic utilizes Gmsh for meshing technically, this means you can use all the options Gmsh offers. Right now -cl is just one global option, but if you would like to have more control over the mesh all you need to do is interfere with meshing step of Mosaic. At the end of the day Mosaic creates a geometry file, .geo_unrolled after that you can decide on element size, element number, element order, element type yourself.) Regarding orientation, can you explain with more details, I am not so sure I understood what you meant here? If you mean a stat file option, maybe you can tell the specific key?

2. Can you open an issue in [our Github ](https://github.com/InstituteOfMechanics/Neper_Mosaic_RVE_Generator/tree/main)please for this one, and there can you post your inp file, do you have ELSET definitions?

Best regards
Dilek

**Reply by MROHOMAN** [reply]
Hi,
Thank you so much for your detailed answer. 

1. By orientation, I meant the Euler angles for each element. NEPER has a command like stateIt which can generate Euler angles for elements. Can you utilize something like that in MOSAIC?

2. I will try to do that.

**Reply by dilekguzel** [reply]
Hey @MROHOMAN,

1. No we don't have this information right now, however one can utilise it by using Gmsh API again.
2. You need to import as a model not as a part. File > Import > Model, and then you will have grain structures.

Best regards
Dilek
